bpo-24744: Raises error in pkgutil.walk_packages if path is str

CuriousLearner opened the pull request:
Attempts to resolve: http://bugs.python.org/issue24744

In the Bug @bitdancer suggested that *some* sort of error should be raised if path is of type str in `pkgutil.walk_packages` function.

@bitdancer For the meanwhile I've raised TypeError since that was looking most appropriate to me. Please let me know what kind of error is expected to be raised here and I'll update the PR. 

Thanks!

CuriousLearner committed:
bpo-24744: Raises error in pkgutil.walk_packages if path is str

bitdancer commented:
I think ValueError would be better, but it doesn't matter all that much (the two have a very blurry line between them in Python).

A test is needed as well.  And we should also add a test that a bytes object also produces some sort of error (we don't care what sort).

CuriousLearner commented:
@bitdancer what would be the most appropriate place to put the test at?

I can see that there is a test written for `pkgutil.walk_packages` in `Lib/test/test_runner.py` by @ncoghlan . Would that file be a good place to add this test?

bitdancer commented:
I haven't reviewed the reason for the 'hack' of putting the walk_packages test in test_runner, but since this is only checking the raising of an error on bad input, I'm guessing you can put it in test_pkgutil.  Try it :)  The long term goal is to move the tests from test_runner to test_pkgutil, so any new tests that *can* go in test_pkgutil should go there.

CuriousLearner commented:
Hi @bitdancer, I've updated the branch with changes and also added a test.

However, I'm not sure why the PR is not reflecting it, although it's from the same branch (https://github.com/CuriousLearner/cpython/tree/fix-issue24744) I'm doing the changes in. This seems weird to me.

CuriousLearner committed:
bpo-24744: Add bad input test for pkgutil.walk_packages() function

CuriousLearner commented:
Never mind, I figured that out, it was a minor naming issue with the branch. The patch would be now updated here.

Please have a look @bitdancer . Let me know if this needs changes.

Thanks!

ncoghlan commented:
@bitdancer Some of the tests currently in test_runpy are there not because that's the right place for them, but because they re-use some of the testing machinery that's currently in that file. There are a couple of old issues about moving those helpers out into the `test.support` subpackage, allowing the affected test cases to be moved to `test_pkgutil` where they belong:

* Moving the walk_packages helpers out of `test_runpy`: https://bugs.python.org/issue15376
* Moving the sample package creation helpers to a common location: https://bugs.python.org/issue15403

bitdancer commented on Lib/test/test_pkgutil.py:
```diff
@@ -176,6 +176,21 @@ def test_walkpackages_zipfile(self):
+        with self.assertRaises(ValueError) as context:
+            list(pkgutil.walk_packages(str_input))
+
+        self.assertTrue('path must be None or list of paths' in str(context.exception))
```
I don't think this assertion is needed.  All we care about here is that the ValueError gets raised, we really don't care what the error message is.  (This is not always true, but in this case I think it is.)

In fact, in this particular case we don't even care whether we get a ValueError or a TypeError, so I'd make the test accept both as passing.  That way we're doing the same check for both str and bytes: all we are trying to prove with these tests is that we get an error rather than an empty list.

bitdancer commented on Lib/pkgutil.py:
```diff
@@ -119,6 +119,9 @@ def iter_modules(path=None, prefix=''):
     """
     if path is None:
         importers = iter_importers()
+    elif isinstance(path, str) or isinstance(path, bytes):
```
isinstance(path, bytes) is problematic.  I was hoping that walk_packages would throw some other error for bytes input (perhaps a TypeError on some string operation), because checking for all the bytes types is a bit of a pain.  And testing it, this is in fact true.  So the bytes check isn't needed, just the str check.  (Yes, the error is a bit mysterious, but if we want to address that, it requires more of a design discussion, and we shouldn't deal with that in this particular issue).

CuriousLearner replied:
Hi @bitdancer , what is meant by:

> checking for all the bytes types

bitdancer replied:
bytes, bytearray, array('b') (or whatever the right incantation is), memoryview....

CuriousLearner committed:
bpo-24744: Address review comments

bitdancer commented on Lib/test/test_pkgutil.py:
```diff
@@ -181,17 +181,13 @@ def test_walkpackages_on_bad_input(self):
-
         bytes_input = b'test_dir'
-        with self.assertRaises(ValueError) as context:
+        with self.assertRaises(TypeError):
```
I would like both of these to be `assertRaises((TypeError, ValueError))`.

CuriousLearner replied:
@bitdancer Fixed. Does this looks good now?

CuriousLearner committed:
bpo-24744: Address review comments in pkgutil tests

CuriousLearner committed:
bpo-24744: Fixes docstring of test_walkpackages_on_bad_input method

bitdancer commented:
It would be better if the two tuples were in the same order, but I wouldn't have bothered you with that except that I'd like to request that you delete the test method docstring.  We don't in general use docstrings for test methods in the cpython test suite.  If there is important information not included in the test name, you can use a comment.  But I'd just rename the test "test_walk_packages_raises_on_string_or_bytes_input".

CuriousLearner committed:
bpo-24744: Address review comments

CuriousLearner commented:
Sure @bitdancer , I've done the changes :)

CuriousLearner committed:
bpo-24744: Adds entry in Whatsnew/3.7.rst and NEWS section

bitdancer commented on Doc/whatsnew/3.7.rst:
```diff
@@ -330,6 +330,11 @@ that may require changes to your code.
+* :meth:`pkgutil.walk_packages` now raises ValueError/TypeError if path is
+  string/bytes object.
+  Previously an empty list was returned. (Contributed by Sanyam Khurana in
+  :issue:`24744`.)
```
Please reflow this into a single paragraph.  And it should only mention str; it has always raised a TypeError for bytes.

bitdancer reviewed (changes requested):
I don't see the NEWS entry, I suspect you forgot to 'git add' it.

CuriousLearner committed:
bpo-24744: Adds entry in Whatsnew/3.7.rst and NEWS section

CuriousLearner committed:
bpo-24744: Minor fix in Misc/NEWS

bitdancer commented on Doc/whatsnew/3.7.rst:
```diff
@@ -330,10 +330,9 @@ that may require changes to your code.
-  :issue:`24744`.)
+* :meth:`pkgutil.walk_packages` now raises ValueError if path is a string
+  object. Previously an empty list was returned. (Contributed by Sanyam
+  Khurana in :issue:`24744`.)
```
'object' is redundant here.  Also, mark up "path" as an argument by surrounding it with asterisks.

CuriousLearner replied:
FIxed.

bitdancer commented on Misc/NEWS:
```diff
@@ -345,9 +345,11 @@ Extension Modules
 -------
 
+- bpo-24744: pkgutil.walk_packages function raises ValueError if path is of
+  str object type. Patch by Sanyam Khurana.
```
... `function now raises a ValueError if *path* is a string` ...

CuriousLearner replied:
Fixed.

CuriousLearner committed:
bpo-24744: Minor fix in NEWS and whatsnew/3.7.rst

zware committed:
Merge branch 'master' into fix-Issue24744

CuriousLearner committed:
bpo-24744: Merge branch 'master' & resolves merge conflicts

bitdancer committed:
Merge branch 'master' into fix-Issue24744

CuriousLearner commented:
Thanks @bitdancer :)
